### PR TITLE
Reduce str transmutes, add mut versions of methods.

### PR DESCRIPTION
**Pull request by clarfonthey**
When I was working on the various parts involved in #40380 one of the comments I got was the excess of transmutes necessary to make the changes work. This is part of a set of multiple changes I'd like to offer to fix this problem.

I think that having these methods is reasonable because they're already possible via transmutes, and it makes the code that uses them safer. I can also add `pub(crate)` to these methods for now if the libs team would rather not expose them to the public without an RFC.

### CODE REVIEW

File: src/libcollections/str.rs [hunk 1]
```diff
@@ -294,6 +294,13 @@ impl str {
         core_str::StrExt::as_bytes(self)
     }
 
+    /// Converts a mutable string slice to a mutable byte slice.
+    #[unstable(feature = "str_mut_extras", issue = "0")]
+    #[inline(always)]
+    pub fn as_bytes_mut(&mut self) -> &mut [u8] {
```
Comment (sfackler) [inline src/libcollections/str.rs]: This would have to be unsafe.

Comment (clarcharr) [inline src/libcollections/str.rs]: For some reason I thought I annotated this as unsafe but I didn't.

**Comment by alexcrichton** [comment]
These seem like reasonable additions to me! 

(clear mirrors of the slice equivalents)

**Comment by alexcrichton** [comment]
Looks like travis also has failures?

**Comment by clarcharr** [comment]
Travis failures are due to a requirement for docs of the new features; I will fix those in a bit.

**Comment by clarcharr** [comment]
Also I just added a `from_utf8_mut` as well.

**Comment by alexcrichton** [comment]
Looks like the build may still be failing?

**Comment by clarcharr** [comment]
Seems to be passing now!

**Comment by TimNN** [comment]
@bors retry

* #40546

**Comment by TimNN** [comment]
@bors retry

* #40546

**Comment by TimNN** [comment]
@bors retry

* #40546

**Pull request merged**